pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,243 @@ package heal
+		//case <-time.After(time.Until(expireTime)):
+		case <-time.After(time.Second):
+			u.removeFromExpectedConnections(conn)
+			return conn, errors.New("timeout expired but we couldn't verify that connection was established")
+		}
 	}
-	return conn, err
+
+	return conn, nil
 }
 
 func (u *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	if cachedConn, err := u.removeConnectionFromMonitor(conn); err == nil {
+		conn = cachedConn
+	} else {
+		// TODO log this?
+	}
+
 	return next.Client(ctx).Close(ctx, conn, opts...)
 }
+
+func (u *healClient) listenToConnectionChanges(heal HealRequestFunc, errCh chan error) {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{})
+	if err != nil {
+		errCh <- errors.Wrap(err, "MonitorConnections failed")
+		return
+	}
+	close(errCh)
+
+	// TODO should we really compare paths from the beginning?
+	pathStartsWith := func(c *networkservice.Connection, reference *networkservice.Connection) bool {
+		if uint32(len(c.GetPath().GetPathSegments())) <= reference.GetPath().GetIndex() {
+			return false
+		}
+		for i := uint32(0); i < reference.GetPath().GetIndex(); i++ {
+			if !c.GetPath().PathSegments[i].Equal(reference.GetPath().GetPathSegments()[i]) {
+				return false
+			}
+		}
+		return true
+	}
+
+	for u.ctx.Err() == nil {
+		event, err := monitorClient.Recv()
+		if err != nil {
+			if u.ctx.Err() != nil {
+				return
+			}
+			monitorClient, err = u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{})
+			if err != nil {
+				// TODO can we handle this?
+				return
+			}
+			continue
+		}
+
+		var connsToHeal []string
+
+		switch event.GetType() {
+		case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER: // TODO is this case special?
+			fallthrough
+		case networkservice.ConnectionEventType_UPDATE:
+			// TODO adapt this code, copied from server, to client
+			// If the server has a pathSegment for this Connection.Id, but its not the one we
+			// got back from it... we should fail, as different Request came after ours successfully
+			//if !pathSegment.Equal(conns[pathSegment.GetId()].GetCurrentPathSegment()) {
+			//	return errors.Errorf("server has a different pathSegment than was returned to this call.")
+			//}
+
+			// todo fix this mutex mess
+			u.connCacheMutex.Lock()
+			for _, conn := range event.GetConnections() {
+				if _, ok := u.connCache[conn.GetId()]; ok {
+					u.connCache[conn.GetId()] = conn.Clone()
+				}
+			}
+			u.connCacheMutex.Unlock()
+
+			u.expectedConnsMutex.Lock()
+			for _, conn := range event.GetConnections() {
+				for _, svReq := range u.expectedConns {
+					if pathStartsWith(conn, svReq.prefixConn) {
+						svReq.ch <- conn.GetPath().GetPathSegments()[svReq.prefixConn.GetPath().GetIndex()+1]
+					}
+				}
+			}
+			u.expectedConnsMutex.Unlock()
+
+		case networkservice.ConnectionEventType_DELETE:
+			if heal != nil {
```

Comment:
It must be not nil, so you may add this check somewhere out of this function.

Reply:
When I talked with Denis Tingaikin about heal callback few days ago, he said that heal server may not be present in the chain, therefore it's allowed for this variable to be nil, and heal client in this case should still listen to changes to replace the connection in the Close() call.